Purchases/Purchasing/ProductDetails.swift

```diff
@@ -0,0 +1,137 @@
+    // https://github.com/RevenueCat/purchases-ios/issues/848
+    //    @available(iOS 11.2, *)
+    //    var introductoryPrice: SKProductDiscount? { get }
+    //    //
+    //    @available(iOS 12.2, *)
+    //    var discounts: [SKProductDiscount] { get }
+}
+
+@available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+@objc(RCSK2ProductDetails) public class SK2ProductDetails: ProductDetails {
+
+    init(sk2Product: SK2Product) {
+        self._underlyingSK2Product = sk2Product
+    }
+
+    // We can't directly store instances of StoreKit.Product, since that causes
+    // linking issues in iOS < 15, even with @available checks correctly in place.
+    // So instead, we store the underlying product as Any and wrap it with casting.
+    // https://openradar.appspot.com/radar?id=4970535809187840
+    private var _underlyingSK2Product: Any
+    public var underlyingSK2Product: SK2Product {
+        // swiftlint:disable:next force_cast
+        get { _underlyingSK2Product as! SK2Product }
+        set { _underlyingSK2Product = newValue }
+    }
+
+    @objc public override var localizedDescription: String { underlyingSK2Product.description }
+
+    @objc public override var price: Decimal { underlyingSK2Product.price }
+
+    @objc public override var localizedPriceString: String { underlyingSK2Product.displayPrice }
+
+    @objc public override var productIdentifier: String { underlyingSK2Product.id }
+
+    @objc public override var isFamilyShareable: Bool { underlyingSK2Product.isFamilyShareable }
+
+    @objc public override var localizedTitle: String { underlyingSK2Product.displayName }
+
+    @objc public override var subscriptionGroupIdentifier: String? {
+        underlyingSK2Product.subscription?.subscriptionGroupID
+    }
+
+}
+
+@objc(RCSK1ProductDetails) public class SK1ProductDetails: ProductDetails {
+
+    @objc public init(sk1Product: SK1Product) {
+        self.underlyingSK1Product = sk1Product
+    }
+
+    @objc public let underlyingSK1Product: SK1Product
+
+    @objc public override var localizedDescription: String { return underlyingSK1Product.localizedDescription }
+
+    @objc public override var price: Decimal { return underlyingSK1Product.price as Decimal }
+
+    @objc public override var localizedPriceString: String {
+        return formatter.string(from: underlyingSK1Product.price) ?? ""
+    }
+
+    @objc public override var productIdentifier: String { return underlyingSK1Product.productIdentifier }
+
+    @available(iOS 14.0, macOS 11.0, tvOS 14.0, watchOS 8.0, *)
+    @objc public override var isFamilyShareable: Bool { underlyingSK1Product.isFamilyShareable }
+
+    @objc public override var localizedTitle: String { underlyingSK1Product.localizedTitle }
+
+    @available(iOS 12.0, macCatalyst 13.0, tvOS 12.0, macOS 10.14, watchOS 6.2, *)
+    override public var subscriptionGroupIdentifier: String? { underlyingSK1Product.subscriptionGroupIdentifier }
+
+    private lazy var formatter: NumberFormatter = {
```

Comment:
probably not, but there's no harm in making this a regular variable. I'm not sure why I made this a lazy var tbh.

Reply:
FWIW they're very slow to initialize so they are worth caching.

Reply:
the bummer here is that they're currently set up to be tied to the instance, because they're getting the locale from the SKProduct. 
So we'd be doing it once per product. 

One probably better alternative is to have it as a static var, and initialize it with the price locale of the first SKProduct received. 
Then, listen for [changes to the storefront](https://developer.apple.com/documentation/storekit/skpaymenttransactionobserver/3242936-paymentqueuedidchangestorefront#) and reset if needed. 
What do you think?

Reply:
That seems like a good refactor, worth filing, but I think this is fine like this for now?

Reply:
https://app.shortcut.com/revenuecat/story/11127/optimize-date-formatters-initialization-for-sk2-products